logr.go

```diff
@@ -392,12 +415,78 @@ func (l Logger) IsZero() bool {
 	return l.sink == nil
 }
 
+// WithContext stores a context in the Logger. If the Logger also has keys set
+// that it is meant to log from a context, then the values for those keys from
+// the given context will be added to all log entries.
+func (l Logger) WithContext(ctx context.Context) Logger {
+	if l.skipSink() {
```

Comment:
I was wondering about that myself. But when I looked at https://github.com/go-logr/logr/pull/153 it said that null Logger and Discard logger were intended to be different (implicit vs explicit discard).

I would prefer to drop that distinction. Do you want to submit the patch above as PR?

Reply:
I can also copy it into the "discard" commit - seems related enough that we don't need two commits+PRs.

Reply:
I created https://github.com/go-logr/logr/pull/166 for this and https://github.com/go-logr/logr/pull/165 for the funcr optimizations.

I prefer to keep the benchmarks in this PR for a bit longer because I might still end up modifying them.